src/hotspot/share/gc/g1/g1FullGCPrepareTask.cpp

```diff
@@ -157,6 +166,29 @@ bool G1FullGCPrepareTask::G1CalculatePointersClosure::should_compact(HeapRegion*
   return live_words <= live_words_threshold;
 }
 
+void G1FullGCPrepareTask::G1CalculatePointersClosure::update_bot(HeapRegion* hr) {
+  HeapWord* limit = hr->top();
+  HeapWord* next_addr = hr->bottom();
+  HeapWord* threshold = hr->initialize_threshold();
+  HeapWord* pre_addr;
+
+  while (next_addr < limit) {
+    Prefetch::write(next_addr, PrefetchScanIntervalInBytes);
+    pre_addr = next_addr;
+    if (_bitmap->is_marked(next_addr)) {
+      oop obj = cast_to_oop(next_addr);
+      next_addr += obj->size();
+    } else {
+      next_addr = _bitmap->get_next_marked_addr(next_addr, limit);
+    }
+
+    if (next_addr > threshold) {
+      threshold = hr->cross_threshold(pre_addr, next_addr);
+    }
+  }
+  assert(next_addr == limit, "Should stop the scan at the limit.");
+}
```

Comment:
I think this could be implemented using `apply_to_marked_objects()`. Let the closure keep track of the threshold for the region and the `apply()` function could look something like this:
```
size_t apply(oop object) {
  size_t size = object->size();
  HeapWord* addr = cast_from_oop<HeapWord*>(object);
  HeapWord* next_addr = addr + size;
  if (next_addr > _threshold) {
    _threshold = hr->cross_threshold(addr, next_addr);
  }
  return size;
}
```

Reply:
I realize that we need to keep the previous address in the closure as well, for the case when we step over the threshold because of an unmarked object.

Reply:
Hi Stefan, Thanks for the suggestion. 
At first thought, I think it's a good way to reuse the traversal ability of HeapRegion::apply_to_marked_objects.

But when I try to implement it actually, seems to me it might make the code a little complicated.

For the first part, in the new G1UpdateBotClosure, we need to indroduce an _pre_addr as you suggested. this part is implemented as below, it's not that complicated, please check the following code snippet:
 void G1FullGCPrepareTask::G1CalculatePointersClosure::update_bot(HeapRegion* hr) {
-  ... ...
-  assert(next_addr == limit, "Should stop the scan at the limit.");
+  G1UpdateBotClosure updateBot(hr);
+  hr->apply_to_marked_objects(_bitmap, &updateBot);
 }
+G1FullGCPrepareTask::G1UpdateBotClosure::G1UpdateBotClosure(HeapRegion* hr) :
+    _hr(hr),
+    _pre_addr(hr->bottom()),
+    _threshold(hr->initialize_threshold()) { }
+
+size_t G1FullGCPrepareTask::G1UpdateBotClosure::apply(oop object) {
+  HeapWord* addr = cast_from_oop<HeapWord*>(object);
+  size_t size = object->size();
+  HeapWord* next_addr = addr + size;
+
+  if(addr > _threshold) {
+    _threshold = _hr->cross_threshold(_pre_addr, addr);
+  }
+  if (next_addr > _threshold) {
+    _threshold = _hr->cross_threshold(addr, next_addr);
+  }
+  _pre_addr = next_addr;
+  return size;
+}
+

But, the above code does not consider the situation: one or several dead objects are at the end of the heap region, at this situation, we needs to update the bot outside of the G1UpdateBotClosure after hr->apply_to_marked_objects(...). I did not implement this part yet, but seems to me it makes the code a little complicated and not that readable, so would like to discuss with you first.

In summary, seems it does not make the code more readable, and will increase the complexity.
How do you think about it?

Reply:
BTW, may I ask what's the format do you use the quote a snippet of code, seems my snippet of code lost all format info(e.g. indent).

Reply:
Easiest is to mark the code and push the "Insert code" button (<>). You can also manually write ``` to start a code block. 

I'll get back to you about the code above later.

Reply:
Thanks Stefan, it works.

Reply:
Looked a bit more at the code now and I agree that it didn't become as neat as I had hoped. Solving the problem you describe with dead objects at the end should be fairly easy in the closure-destructor I guess. Something like:
```
  ~G1UpdateBotClosure() {
    if (_hr->top() > _threshold) {
      _hr->cross_threshold(_pre_addr, _hr->top());
    }
  }
```

And a few comments in the apply method will help readability:
```
  size_t apply(oop object) {
    size_t size = object->size();
    HeapWord* addr = cast_from_oop<HeapWord*>(object);
    HeapWord* next_addr = addr + size;

    // Threshold was crossed by dead object.
    if(addr > _threshold) {
      _threshold = _hr->cross_threshold(_pre_addr, addr);
    }

    // Threshold is crossed by this object.
    if (next_addr > _threshold) {
      _threshold = _hr->cross_threshold(addr, next_addr);
    }

    _pre_addr = next_addr;
    return size;
  }
```

But still it might be better to instead streamline the current approach:
```suggestion
  HeapWord* limit = hr->top();
  HeapWord* next_addr = hr->bottom();
  HeapWord* threshold = hr->initialize_threshold();
  HeapWord* pre_addr;

  while (next_addr < limit) {
    pre_addr = next_addr;
    next_addr = _bitmap->get_next_marked_addr(next_addr + 1, limit);
    
    if (next_addr > threshold) {
      threshold = hr->cross_threshold(pre_addr, next_addr);
    }
  }
  assert(next_addr == limit, "Should stop the scan at the limit.");
}
```
Here I removed the `Prefetch` because we do not need to touch the heap. For this to be true I also removed the small optimization to use `obj->size()`, which is probably not an optimization if we don't need to do anything with the "object".  

After doing this I agree that reusing `apply_to_marked_objects()` is probably not helping.

Reply:
Hi Stefan, I think the refactoring you suggested is great, I just modified as you suggested and the pre-submit tests run successfully.